File: online judge.kaner murillo/python/10261.py
#https://onlinejudge.org/external/102/10261.pdf
#10261

from pprint import pprint
from sys import stdin

def solve(i, n, sum1, sum2, cars, mem):

    ans, key = None, (i, abs(sum1-sum2))

    if key in mem:
        ans = mem[key]
    
    else:

        if i == n:

            ans = [0, [-1]]
        
        else:

            a, b = [0, [-1]], [0, [-1]]

            if sum1 - cars[i] >= 0:
                a = list(solve(i+1, n, sum1-cars[i], sum2, cars, mem))
                a[0]+=1
                a[1] = list(a[1])
                a[1].append(0)

            if sum2 - cars[i] >= 0:
                b = list(solve(i+1, n, sum1, sum2-cars[i], cars, mem))
                b[0]+=1
                b[1] = list(b[1])
                b[1].append(1)

            if a[0] >= b[0]:
                ans = a
            
            else:
                ans = b

        mem[key] = ans
                
    return ans
    
def main():

    cases = int(stdin.readline())
    stdin.readline()

    for i in range(cases):
        
        if i:
            stdin.readline()
            print()

        size = int(stdin.readline())
        cars = []

        size_car = int(stdin.readline())

        while size_car:
            cars.append(size_car/100)
            size_car = int(stdin.readline())
        
        mem = {}
        ans = solve(0, len(cars), size, size, cars, mem)
        print(ans[0])

        for j in range(len(ans[1])-1, 0, -1):
            if not ans[1][j]:
                print("port")
            else:
                print("starboard")    
        

main()

File: online judge.kaner murillo/c++/10261.cpp
//https://onlinejudge.org/external/102/10261.pdf
//10261

#include<cstdio>
#include<iostream>
#include<vector>
#include<map>

using namespace std;

pair<int, vector<int>>dp(int i, int n, int sum1, int sum2, vector<int>&cars, map<pair<int, int>, pair<int, vector<int>>>&mem){

    pair<int, vector<int>>ans;

    if(mem.count({i, abs(sum1-sum2)})){
        ans = mem[{i, abs(sum1-sum2)}];
    }
    else{

        if(i == n){
            ans.first = 0;
            ans.second = {-1};
        }

        else{
            pair<int, vector<int>>a = {0, {-1}};
            pair<int, vector<int>>b = {0, {-1}};

            if(sum1 - cars[i] >= 0){
                a = dp(i+1, n, sum1 - cars[i], sum2, cars, mem);
                a.first+=1;
                a.second.push_back(0);
            }

            if(sum2 - cars[i] >= 0){
                b = dp(i+1, n, sum1, sum2 - cars[i], cars, mem);
                b.first+=1;
                b.second.push_back(1);
            }

            if(a.first >= b.first){
                ans = a;
            }
            else{
                ans = b;
            }
        }
        mem[{i, abs(sum1-sum2)}] = ans;
    }

    return ans;
}

int main(){

    int cases, i, size, sizeCar, car, flag = 0;
    pair<int, vector<int>>ans;
    cin>>cases;
    cin.ignore();

    while(cases--){

        if(flag){
            printf("\n");
            cin.ignore();
        }
        flag = 1;
        vector<int>cars;
        map<pair<int, int>, pair<int, vector<int>>>mem;
        cin>>size;
        size*=100;
        
        while(cin>>sizeCar && sizeCar){
            cars.push_back(sizeCar);
        }
        ans = dp(0, cars.size(), size, size, cars, mem);

        printf("%i\n", ans.first);
        for(i = ans.second.size()-1; i>0; i--){
            if(!ans.second[i]){
                printf("port\n");
            }
            else{
                printf("starboard\n");
            }
        }
    }

    return 0;
}